src/opm/parser/eclipse/EclipseState/Schedule/Well/WellTestState.cpp

```diff
@@ -19,72 +19,95 @@
+        if (well_ptr) {
+            // the well exists already, we just update it with action of closing
+            well_ptr->closed = true;
+            well_ptr->last_test = sim_time;
+        } else
+            this->wells.push_back({well_name, reason, true, sim_time, 0});
     }
 
 
-    void WellTestState::openWell(const std::string& well_name) {
-        wells.erase(std::remove_if(wells.begin(),
-                                   wells.end(),
-                                   [&well_name](const ClosedWell& well) { return (well.name == well_name); }),
-                    wells.end());
+    void WellTestState::openWell(const std::string& well_name, WellTestConfig::Reason reason) {
+
+        WTestWell* well_ptr = getWell(well_name, reason);
+
+        if (well_ptr)
+            well_ptr->closed = false;
+        else
+            throw std::runtime_error("No well named " + well_name + " with close reason "
+                                    + WellTestConfig::reasonToString(reason)
+                                    + " found in WellTestState.");
     }
 
 
     void WellTestState::dropWell(const std::string& well_name, WellTestConfig::Reason reason) {
         wells.erase(std::remove_if(wells.begin(),
                                    wells.end(),
-                                   [&well_name, reason](const ClosedWell& well) { return (well.name == well_name && well.reason == reason); }),
+                                   [&well_name, reason](const WTestWell& well) { return (well.name == well_name && well.reason == reason); }),
                     wells.end());
     }
 
 
-    bool WellTestState::hasWell(const std::string& well_name, WellTestConfig::Reason reason) const {
+    bool WellTestState::hasWellClosed(const std::string& well_name, WellTestConfig::Reason reason) const {
         const auto well_iter = std::find_if(wells.begin(),
                                             wells.end(),
-                                            [&well_name, &reason](const ClosedWell& well)
-                                             {
-                                                return (reason == well.reason && well.name == well_name);
+                                            [&well_name, &reason](const WTestWell& well)
+                                            {
+                                                return (reason == well.reason && well.name == well_name && well.closed);
                                             });
         return (well_iter != wells.end());
     }
 
+
+    WellTestState::WTestWell* WellTestState::getWell(const std::string& well_name, WellTestConfig::Reason reason)
+    {
+        const auto well_iter = std::find_if(wells.begin(), wells.end(), [&well_name, &reason](const WTestWell& well) {
+            return (reason == well.reason && well.name == well_name);
+        });
+
+        return (well_iter == wells.end() ? nullptr : &(*well_iter) );
+    }
+
+
```

Comment:
When you are here - please rename `WellTestState::sizeWells()` &rightarrow; `WellTestState::size()`

Reply:
Sure.

Reply:
We can discuss this one later. There is a sizeCompletions() function there. I have not checked the things related to WTEST of Completions yet. How do you think?

Reply:
Ok - then we leave it.

Reply:
I changed it anyway. Later we can look for a better name with updated understanding when necessary.

Reply:
Github webpage is not responsive. Your comment was not shown up when I commented above.

I recover the function name, while we will look for a better name later still.